Improve View Editor

aivuk opened the pull request:
- fixes #114

Improving the View Editor to validate SQL query and deal with errors from the backend.

aivuk commented on src/components/Editors/Table/Table.tsx:
```diff
@@ -93,6 +94,9 @@ export default function Table(props: TableProps) {
     const columns = []
     for (const [key, field] of Object.entries(fields)) {
       const columnIndex = parseInt(key) + 1
+      // Otherwise the _rowNumber and _rowValid are displayed on the table
```
@roll without that the colum _rowNumber and _rowValid were being displayed and the DataGrid complained because there was two columns _rowNumber

roll replied:
Currently, it's not really important I'm going to fix the Table

aivuk commented on src/components/Controllers/View/store.ts:
```diff
@@ -1,8 +1,9 @@
 import * as React from 'react'
 import * as zustand from 'zustand'
 import create from 'zustand/vanilla'
+import { Parser } from 'node-sql-parser'
```
I'm using this library because of the Typescript support and it's easier to check for errors and referenced tables and fields after your parse a query

aivuk commented on src/components/Editors/View/Query.tsx:
```diff
@@ -13,7 +13,7 @@ export default function Query() {
   const setQuery = useStore((state) => state.setQuery)
   const viewError = useStore((state) => state.viewError)
-  const aceEditor = React.useRef<AceEditor>(null);
+  const aceEditor = React.useRef<AceEditor>(null)
```
this is there to make it able us to refer to the editor instance and be able to use its api to insert new text

roll commented on src/components/Controllers/View/store.ts:
```diff
@@ -32,17 +39,89 @@ export function createStore(props: SqlProps) {
       if (!view) return
-      const { table } = await client.tableQuery({ query: view.query })
-      set({ table })
+      const parser = new Parser()
```
We need to split:
- query validation -> Editors/View (as it's shown now for example of the editor loosing focus we do query validation)
- result validation (from server) -> Controllers/View (showing the errors instead of the table)

roll reviewed (changes requested):
Thanks @aivuk it looks good in-general!

I added a few comments and also I'm having a problem with multiline queries (I click enter to expand the editor but already written lines disappear):

![view](https://user-images.githubusercontent.com/557395/221553018-daaa7911-e7c5-4159-ad71-38d5c131decf.png)

aivuk commented:
> Thanks @aivuk it looks good in-general!
> 
> I added a few comments and also I'm having a problem with multiline queries (I click enter to expand the editor but already written lines disappear):
> 
> ![view](https://user-images.githubusercontent.com/557395/221553018-daaa7911-e7c5-4159-ad71-38d5c131decf.png)

The ACE editor needs a height and I think I left it too short

> Thanks @aivuk it looks good in-general!
> 
> I added a few comments and also I'm having a problem with multiline queries (I click enter to expand the editor but already written lines disappear):
> 
> ![view](https://user-images.githubusercontent.com/557395/221553018-daaa7911-e7c5-4159-ad71-38d5c131decf.png)

The Ace editor needs a height and I left it too short at: https://github.com/frictionlessdata/application/blob/114-improve-view-editor/src/components/Editors/View/Query.tsx#L39

roll commented:
I think it can be like `theme.spacing(24)` or something like this at least

aivuk commented:
> I think it can be like `theme.spacing(24)` or something like this at least

Just did it

aivuk commented:
@roll I fixed the remaining issues and now I'm correctly preventing to send an invalid query to the api.

roll reviewed (approved):
Thanks!

I think it's a good start. We still need to finish some tasks from #114 and then work on the UI
